plutus-chain-index/app/Marconi.hs

```diff
@@ -223,23 +228,64 @@ utxoWorker Coordinator{_barrier} ch path = Utxo.open path (Utxo.Depth 2160) >>=
               offset <- findIndex  (\u -> (u ^. Utxo.slotNo) < slot) events
               Ix.rewind offset index
 
+scriptTxWorker
+  :: Coordinator
+  -> TChan (ChainSyncEvent (BlockInMode CardanoMode))
+  -> FilePath
+  -> IO ()
+scriptTxWorker Coordinator{_barrier} ch path = ScriptTx.open path (ScriptTx.Depth 2160) >>= loop
+  where
+    loop :: ScriptTx.ScriptTxIndex -> IO ()
+    loop index = do
+      signalQSemN _barrier 1
+      event <- atomically $ readTChan ch
+      case event of
+        RollForward (BlockInMode (Block (BlockHeader slotNo _ _) txs) era) _ct -> do
+          withIsCardanoEra era (Ix.insert (toUpdate txs slotNo) index >>= loop)
+        RollBackward cp _ct -> do
+          events <- Ix.getEvents (index ^. Ix.storage)
+          loop $
+            fromMaybe index $ do
+              slot   <- chainPointToSlotNo cp
+              offset <- findIndex  (\u -> ScriptTx.slotNo u < slot) events
+              Ix.rewind offset index
+
+    toUpdate :: forall era . C.IsCardanoEra era => [Tx era] -> SlotNo -> ScriptTx.ScriptTxUpdate
+    toUpdate txs slotNo = ScriptTx.ScriptTxUpdate txScripts' slotNo
+      where
+        txScripts' = map (\tx -> (txCbor tx, txScripts tx)) txs
+
+    txCbor :: forall era . C.IsCardanoEra era => Tx era -> ScriptTx.TxCbor
+    txCbor tx = ScriptTx.TxCbor $ C.serialiseToCBOR tx
+
+    txScripts :: forall era . Tx era -> [ScriptTx.ScriptAddress]
+    txScripts tx = let
+        Tx (body :: C.TxBody era) _ws = tx
+        map' = plutusScriptsFromTxBody body :: M.Map Ledger.ScriptHash Ledger.Script
```

Comment:
I don't suggest using `plutusScriptsFromTxBody`. Marconi is an off-chain component, so it should be using types for off-chain use (like `cardano-api`). Anything with with `plutus-ledger-api` types are really meant for on-chain use.

I suggest reimplementing it in Marconi (something named like `getScriptHashesFromTx`) and write tests for it.

Reply:
`plutusScriptsFromTxBody` hashes the script by
- deserialising it into a plutus-ledger `Script` (defined in Plutus.V1.Ledger.Scripts)
- scriptHash serialises it again to bytes and makes a cardano-api
  `Script` (defined in Cardano.Api.Script; that script wraps the
  serialised bytes), then hashes that script

If the goal is to skip plutus-ledger `Script` then is it ok if I wrap the received bytes into cardano-api's `Script` directly? I do this [here](https://github.com/input-output-hk/plutus-apps/blob/d376c2835deec98f8f8475ee9a0e50472c3cefd4/plutus-chain-index/app/Marconi.hs#L277-L287), specifically within the `mkCardanoApiScript` which wraps the incoming bytes into a cardano-api `Script` (the previous way to get this script included going through deserialisation/serialisation steps).

Otherwise, I haven't found a way to go from the bytes I receive to a cardano-api `Script` without using anything from the ledger.

---

Regarding tests, since it's a property test, then if I generate arbitrary scripts, serialize them and get a hash, and then put the scripts into a transaction to see if I get the same scripts (= same hashes) out again, then what would it be testing? (they are the same by definition) Or, perhaps there are multiple locations to put those scripts and I'd see if I lose any by not handling a branch?

Reply:
Hmm... ideally we would use types from `cardano-api`, not `cardano-ledger`. I recommend not pattern matching on the body of the transaction. I recommend pattern matching with `pattern TxBody :: TxBodyContent ViewTx era -> TxBody era` in `Cardano.Api.TxBody`.

There are some examples in `Ledger.Tx.CardanoAPI`.

Regarding property testing, here's a starting point. 
You can test the `txScripts` function as follows. Generate a random number of plutus scripts (see `genPlutusScript` from `Gen.Cardano.Api.Typed`), then generate an `TxIns` which spends funds from these scripts, then generate a transaction with this `TxIns` (based on `genTxBodyContent`), run your `txScripts` on this generated tx, and verify that the initial generated plutus scripts are part of the function's output.

Reply:
It seems that the scripts I'm picking out and using are not included in the pattern TxBody: https://github.com/input-output-hk/cardano-node/blob/master/cardano-api/src/Cardano/Api/TxBody.hs#L2070

Question is, if the scripts that I pick out are the right ones? As I found also that inside TxBody the Script type also exist under:
- TxBodyContent{txOuts} -> TxOut -> ReferenceScript -> ScriptInAnyLang
- TxBodyContent{txReturnCollateral} -> TxReturnCollateral -> TxOut
- TxBodyContent{txAuxScripts} -> TxAuxScripts -> ScriptInEra

I would guess the scripts at the inputs (the ones that I'm using already and that come in as serialised) are the important ones, as those are the ones that are run.

Reply:
Another idea is that maybe plutus-streaming library (the one that provides the `withChainSyncEventStream` that provides the blocks) can convert scripts to the correct type such that only downstream using cardano-api is sufficient.

Reply:
@koslambrou Have managed to come up with this property test:
https://github.com/input-output-hk/plutus-apps/blob/edc8da432275c017ceba3a3aea24bbe9cb8e5e4b/plutus-chain-index-core/test/MarconiSpec.hs#L26-L66

There is still a piece missing: what does it mean to "spend a script that I created"? In the above I've tried to generate the same number of TxIns as there are scripts and replace the scripts' hashes into the `TxId` field of `TxIn` -- is this the correct way to think about "spending a script"?